frame/nfts/src/lib.rs

```diff
@@ -1578,5 +1624,92 @@ pub mod pallet {
+
+		/// Register a new atomic swap, declaring an intention to send an `item` in exchange for
+		/// `desired_item` from origin to target on the current blockchain.
+		/// The target can execute the swap during the specified `duration` of blocks (if set).
+		/// Additionally, the price could be set for the desired `item`.
+		///
+		/// Origin must be Signed and must be an owner of the `item`.
+		///
+		/// - `collection`: The collection of the item.
+		/// - `item`: The item an owner wants to give.
+		/// - `desired_collection`: The collection of the desired item.
+		/// - `desired_item`: The desired item an owner wants to receive.
+		/// - `maybe_price`: The price an owner is willing to pay for the desired `item`.
+		/// - `maybe_duration`: Optional deadline for the swap. Specified by providing the
+		/// 	number of blocks after which the swap will expire.
+		///
+		/// Emits `SwapCreated` on success.
+		#[pallet::weight(T::WeightInfo::create_swap())]
+		pub fn create_swap(
+			origin: OriginFor<T>,
+			collection: T::CollectionId,
+			item: T::ItemId,
+			desired_collection: T::CollectionId,
+			desired_item: T::ItemId,
+			maybe_price: Option<ItemPrice<T, I>>,
+			maybe_duration: Option<<T as SystemConfig>::BlockNumber>,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			Self::do_create_swap(
+				origin,
+				collection,
+				item,
+				desired_collection,
+				desired_item,
+				maybe_price,
+				maybe_duration,
+			)
+		}
+
+		/// Cancel an atomic swap.
+		///
+		/// Origin must be Signed and must be an owner of the `item` if the deadline hasn't expired.
+		///
+		/// - `collection`: The collection of the item.
+		/// - `item`: The item an owner wants to give.
+		///
+		/// Emits `SwapCancelled` on success.
+		#[pallet::weight(T::WeightInfo::cancel_swap())]
+		pub fn cancel_swap(
+			origin: OriginFor<T>,
+			collection: T::CollectionId,
+			item: T::ItemId,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			Self::do_cancel_swap(origin, collection, item)
+		}
+
+		/// Claim an atomic swap.
+		///
+		/// Origin must be Signed and must be an owner of the `item`.
+		///
+		/// - `send_collection`: The collection of the item to be sent.
+		/// - `send_item`: The item to be sent.
+		/// - `receive_collection`: The collection of the item to be received.
+		/// - `receive_item`: The item to be received.
+		/// - `maybe_receive_amount`: An optional amount to be received.
+		///
+		/// Emits `SwapClaimed` on success.
+		#[pallet::weight(T::WeightInfo::claim_swap())]
+		pub fn claim_swap(
+			origin: OriginFor<T>,
+			send_collection: T::CollectionId,
+			send_item: T::ItemId,
+			receive_collection: T::CollectionId,
+			receive_item: T::ItemId,
+			maybe_receive_amount: Option<ItemPrice<T, I>>,
+		) -> DispatchResult {
+			let origin = ensure_signed(origin)?;
+			Self::do_claim_swap(
+				origin,
+				send_collection,
+				send_item,
+				receive_collection,
+				receive_item,
+				maybe_receive_amount,
+			)
```

Comment:
do_* functions has nearly the same format as top-level methods. The main benefit is that now those do_* functions could be called from the other pallets, so we make our pallets much more re-usable and flexible.
Soon, I'll open a PR where I move all the Uniques call methods into `do_*` sub-functions

Reply:
Correct me if I'm wrong - we only needs the `do_..` helper because of the origin check?
Can't help but wonder if we should eventually come up with a better solution for this.

Reply:
`do_*` could also be called externally. It's a common use case when other projects create a wrapper over our pallet, so extracting the main logic into the `do_*` method makes the whole pallet more flexible